File: 3600-maximize-spanning-tree-stability-with-upgrades/3600-maximize-spanning-tree-stability-with-upgrades.cpp
class Solution {
public:
    int components;
    vector<int> parent, size;

    int Union(int u, int v){
        int pu = findParent(u);
        int pv = findParent(v);

        if(pu == pv) return 0;

        components--;
        if(size[pu] > size[pv]){
            size[pu] += size[pv];
            parent[pv] = pu;
        }
        else{
            size[pv] += size[pu];
            parent[pu] = pv;
        }
        return 1;
    }

    int findParent(int node){
        if(parent[node] == node) return node;
        return parent[node] = findParent(parent[node]);
    }

    int maxStability(int n, vector<vector<int>>& edges, int k) {
        components = n;
        parent.resize(n);
        size.resize(n, 1);

        for(int i = 0; i < n; i++){
            parent[i] = i;
        }

        vector<vector<int>> must, flex;

        for(auto &it : edges){
            if(it[3] == 1) must.push_back(it);
            else flex.push_back(it);
        }

        int mini = INT_MAX;

        for(auto &it : must){
            int u = it[0];
            int v = it[1];
            int w = it[2];

            mini = min(mini, w);

            if(!Union(u, v)){
                return -1;
            }
        }

        sort(flex.begin(), flex.end(), [](vector<int>& a, vector<int>& b){
            return a[2] > b[2];
        });

        priority_queue<int, vector<int>, greater<int>> pq;

        for(auto &it : flex){
            int u = it[0];
            int v = it[1];
            int w = it[2];

            if(Union(u, v)){
                pq.push(w);
            }
        }

        while(k-- && !pq.empty()){
            int x = pq.top();
            pq.pop();
            mini = min(mini, 2 * x);
        }

        if(components != 1) return -1;

        if(!pq.empty()){
            return min(mini, pq.top());
        }

        return mini;
    }
};

File: 3600-maximize-spanning-tree-stability-with-upgrades/3600-maximize-spanning-tree-stability-with-upgrades.java
class Solution {
    int components;
    int[] parent;
    int[] size;
    int findParent(int node){
        if(parent[node] == node) return node;
        return parent[node] = findParent(parent[node]);
    }
    boolean union(int u, int v){
        int pu = findParent(u);
        int pv = findParent(v);

        if(pu == pv) return false;

        components--;

        if(size[pu] > size[pv]){
            parent[pv] = pu;
            size[pu] += size[pv];
        }else{
            parent[pu] = pv;
            size[pv] += size[pu];
        }

        return true;
    }

    public int maxStability(int n, int[][] edges, int k) {

        components = n;
        parent = new int[n];
        size = new int[n];

        for(int i = 0; i < n; i++){
            parent[i] = i;
            size[i] = 1;
        }

        List<int[]> must = new ArrayList<>();
        List<int[]> flex = new ArrayList<>();

        for(int[] e : edges){
            if(e[3] == 1) must.add(e);
            else flex.add(e);
        }

        int mini = Integer.MAX_VALUE;

        for(int[] e : must){
            int u = e[0];
            int v = e[1];
            int w = e[2];

            mini = Math.min(mini, w);

            if(!union(u,v)) return -1;
        }
        flex.sort((a,b) -> b[2] - a[2]);
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(int[] e : flex){
            int u = e[0];
            int v = e[1];
            int w = e[2];

            if(union(u,v)){
                pq.add(w);
            }
        }
        while(k-- > 0 && !pq.isEmpty()){
            int x = pq.poll();
            mini = Math.min(mini, 2 * x);
        }
        if(components != 1) return -1;
        if(!pq.isEmpty()){
            return Math.min(mini, pq.peek());
        }
        return mini;
    }
}